### PR TITLE
Fix intra-doc links for non-path primitives

### LINKED ISSUE #63351: Intra-rustdoc links do not work for all primitives
Coming from the tracking issue #43466.

I'm trying to link to [`slice::rotate_left`](https://doc.rust-lang.org/std/primitive.slice.html#method.rotate_left) and it doesn't seem to be supported I tested [`pointer::is_null`](https://doc.rust-lang.org/stable/std/primitive.pointer.html#method.is_null) as well, to test if it's just polymorphic primitives, and it seems that might be the issue, since it also doesn't work as a intra-link.

/cc @GuillaumeGomez

### PR DESCRIPTION
**Pull request by jyn514**
This does *not* currently work for associated items that are
auto-implemented by the compiler (e.g. `never::eq`), because they aren't
present in the source code. I plan to fix this in a follow-up PR.

Fixes https://github.com/rust-lang/rust/issues/63351 using the approach mentioned in https://github.com/rust-lang/rust/issues/63351#issuecomment-683352130.

r? @Manishearth 

cc @petrochenkov - this makes `rustc_resolve::Res` public, is that ok? I'd just add an identical type alias in rustdoc if not, which seems a waste.

### CODE REVIEW

**Comment by jyn514** [comment]
Another bug I found while converting the standard library: `!` doesn't work, the `!` gets stripped somewhere (maybe in `Disambiguator::from_str`?)

**Comment by jyn514** [comment]
I think I went a little overboard with the symbols ... I'd be ok with removing most of them. I do want to keep `()` and `!` though, I expect people to try and use them.

**Comment by jyn514** [comment]
> I think the issue of linking to primitive types like slice could be fixed before stabilization. How about adding a new "namespace" primitive@ to allow linking to all primitives? For example primitive@slice would link to the slice page. This would allow removing the workaround in #74078.

@ollie27 I finally got this working :)

File: compiler/rustc_resolve/src/lib.rs [hunk 1]
```diff
@@ -67,7 +67,7 @@ use imports::{Import, ImportKind, ImportResolver, NameResolution};
 use late::{HasGenericParams, PathSource, Rib, RibKind::*};
 use macros::{MacroRulesBinding, MacroRulesScope, MacroRulesScopeRef};
 
-type Res = def::Res<NodeId>;
+pub type Res = def::Res<NodeId>;
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/lib.rs:70]: This is more like a local import for a commonly used name.
It's better to add a separate one to rustdoc.

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 2]
```diff
@@ -2003,53 +2053,48 @@ fn handle_variant(
...
+        "bool" | "true" | "false" => Bool,
+        "str" => Str,
+        "slice" | "&[]" | "[T]" => Slice,
+        "array" | "[]" | "[T;N]" => Array,
+        "tuple" | "(,)" => Tuple,
+        "unit" | "()" => Unit,
+        "pointer" | "*" | "*const" | "*mut" => RawPointer,
+        "reference" | "&" | "&mut" => Reference,
+        "fn" => Fn,
+        "never" | "!" => Never,
```
Comment (petrochenkov) [inline src/librustdoc/passes/collect_intra_doc_links.rs:2093]: My only hope is that people won't start asking to add all this stuff to the language after seeing it in docs.

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 3]
```diff
@@ -1018,7 +1018,7 @@ impl LinkCollector<'_, '_> {
             (link.trim(), None)
         };
 
-        if path_str.contains(|ch: char| !(ch.is_alphanumeric() || ":_<>, !".contains(ch))) {
+        if path_str.contains(|ch: char| !(ch.is_alphanumeric() || ":_<>, !*()[]&;".contains(ch))) {
```
Comment (Manishearth) [inline src/librustdoc/passes/collect_intra_doc_links.rs:1021]: Won't this make us bail for tuples?

**Review by Manishearth — APPROVED** [review]

File: src/librustdoc/passes/collect_intra_doc_links.rs (hunk 3 above)
Comment (jyn514) [reply src/librustdoc/passes/collect_intra_doc_links.rs:1021]: What do you mean? This says "ignore the link unless it matches `a-z` or one of the whietlisted characters; this is explicitly adding `()` to the whitelist. Before, `()` would make it ignore the link.

Comment (Manishearth) [reply src/librustdoc/passes/collect_intra_doc_links.rs:1023]: oh, i'm dumb.

Comment (Manishearth) [reply src/librustdoc/passes/collect_intra_doc_links.rs:1023]: r=me once you think petrochenkov's review is satisfied!

Comment (jyn514) [reply src/librustdoc/passes/collect_intra_doc_links.rs:1023]: It's a sign this logic is too complicated and we need to simplify it :P

**Comment by jyn514** [comment]
[Failed in rollup](https://github.com/rust-lang/rust/pull/80343#issuecomment-750678905):

```
 Documenting rustc_feature v0.0.0 (/checkout/compiler/rustc_feature)
[RUSTC-TIMING] rustc_parse_format test:false 0.192
[RUSTC-TIMING] rustc_feature test:false 0.383
error: unresolved link to `repr(transparent)`
   --> compiler/rustc_feature/src/active.rs:46:19
    |
46  |                   $(#[doc = $doc])*
...
...
95  | / declare_features! (
96  | |     // -------------------------------------------------------------------------
97  | |     // feature-group-start: internal feature gates
98  | |     // -------------------------------------------------------------------------
628 | |     // -------------------------------------------------------------------------
629 | | );
    | |__- in this macro invocation
    |
    |
    = note: `-D broken-intra-doc-links` implied by `-D warnings`
    = note: the link appears in this line:
            
            Allows #[repr(transparent)] on unions (RFC 2645).
                     ^^^^^^^^^^^^^^^^^
    = note: no item named `repr(transparent)` in scope
    = help: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`
    = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
```
https://github.com/rust-lang/rust/blob/c34c015fe2710caf53ba7ae9d1644f9ba65a6f74/compiler/rustc_feature/src/active.rs#L476-L477

So we've finally run into a false positive like @Manishearth was worried about. There are a few possible approaches:
1. Say it's user error and `#[repr(transparent)]` should be in backticks to mark it as a code block
2. Remove some subset of the new symbols (at least `[]` or `()`, possibly both).
3. Remove the symbols altogether and only allow using the words (`pointer`, `reference`).

I don't like 1; it seems like busywork for the user. I don't like 3 because it means you can't link to `!` (most people don't know it means 'never' I think). So that only leaves 2 and I don't like that much either because `()` and `[]` both seem like reasonable symbols :/

I guess there's a third option which is to special case `()`, `!`, and `[]` and never involve the rustc_resolve machinery for them at all? Then anything else with those symbols would be ignored by before.

@Manishearth what do you think?

**Comment by Manishearth** [comment]
> Remove some subset of the new symbols (at least [] or (), possibly both).

This is my preferred option. I do not think intra-doc links should be trying to handle any Rust you can throw at it.

**Comment by jyn514** [comment]
@Manishearth do you want to take another look?

**Review by Manishearth — APPROVED** [review]

**Comment by pthariensflame** [comment]
Does this need `relnotes` (for Rustdoc)?

**Comment by jyn514** [comment]
Probably https://github.com/rust-lang/rust/pull/74489 and https://github.com/rust-lang/rust/pull/76934 should have relnotes too then 🤷

**Comment by XAMPPRocky** [comment]
Rustbot added this to the 1.51.0 milestone, but isn't this actually in 1.50.0 because the 27th of December was before the beta cutoff?

**Comment by jyn514** [comment]
@XAMPPRocky I don't see it in the git history for the beta branch. @Mark-Simulacrum may know more but I'd be inclined to trust rustbot.

**Comment by Mark-Simulacrum** [comment]
Beta cut off, per the release process, is when the version number bump lands (https://forge.rust-lang.org/release/process.html#bump-the-stable-version-number-t-6-days-friday-the-week-before), happening on friday usually before the release. Triagebot is correct in this case that this PR will be released in 1.51.

**Pull request merged**